staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go

```diff
@@ -0,0 +1,370 @@
+	for gv := range h.excludedGVs {
+		excludedCloned[gv] = struct{}{}
+	}
+	h.excludedGVsMu.RUnlock()
+
+	filtered := make(map[string]PeerDiscoveryCacheEntry, len(cacheMap))
+	peerDiscoveryChanged := false
+	for peerID, entry := range cacheMap {
+		newEntry, peerChanged := h.filterPeerCacheEntry(entry, excludedCloned)
+		filtered[peerID] = newEntry
+		if peerChanged {
+			peerDiscoveryChanged = true
+		}
+	}
+
+	return filtered, peerDiscoveryChanged
+}
+
+// filterPeerCacheEntry filters a single peer's cache entry for excluded groups.
+// Returns the filtered entry and whether any excluded group was present.
+func (h *peerProxyHandler) filterPeerCacheEntry(entry PeerDiscoveryCacheEntry, excluded map[schema.GroupVersion]struct{}) (PeerDiscoveryCacheEntry, bool) {
+	filteredGVRs := make(map[schema.GroupVersionResource]bool, len(entry.GVRs))
+	peerDiscoveryChanged := false
+
+	for existingGVR, v := range entry.GVRs {
+		gv := schema.GroupVersion{Group: existingGVR.Group, Version: existingGVR.Version}
+		if _, skip := excluded[gv]; skip {
+			peerDiscoveryChanged = true
+			continue
+		}
+		filteredGVRs[existingGVR] = v
+	}
+
+	if !peerDiscoveryChanged {
+		return entry, false
+	}
+
+	// Filter the group discovery list
+	filteredGroups := h.filterGroupDiscovery(entry.GroupDiscovery, excluded)
+	return PeerDiscoveryCacheEntry{
+		GVRs:           filteredGVRs,
+		GroupDiscovery: filteredGroups,
+	}, true
+}
+
+func (h *peerProxyHandler) filterGroupDiscovery(groupDiscoveries []apidiscoveryv2.APIGroupDiscovery, excluded map[schema.GroupVersion]struct{}) []apidiscoveryv2.APIGroupDiscovery {
+	var filteredDiscovery []apidiscoveryv2.APIGroupDiscovery
+	for _, groupDiscovery := range groupDiscoveries {
+		filteredGroup := apidiscoveryv2.APIGroupDiscovery{
+			ObjectMeta: groupDiscovery.ObjectMeta,
+		}
+
+		for _, version := range groupDiscovery.Versions {
+			gv := schema.GroupVersion{Group: groupDiscovery.Name, Version: version.Version}
+			if _, found := excluded[gv]; found {
+				// This version is excluded, skip it
+				continue
+			}
+			filteredGroup.Versions = append(filteredGroup.Versions, version)
+		}
+
+		// Only add the group to the final list if it still has any versions left
+		if len(filteredGroup.Versions) > 0 {
+			filteredDiscovery = append(filteredDiscovery, filteredGroup)
+		}
+	}
+	return filteredDiscovery
+}
+
+// RunGVDeletionWorkers starts worker goroutines to process the GV deletion queue.
+func (h *peerProxyHandler) RunGVDeletionWorkers(ctx context.Context, workers int) {
+	defer h.gvDeletionQueue.ShutDown()
+
+	klog.Infof("Starting %d GV deletion worker(s)", workers)
+	for i := 0; i < workers; i++ {
+		go func(workerID int) {
+			for h.processNextGVDeletion() {
+				select {
+				case <-ctx.Done():
+					klog.Infof("GV deletion worker %d shutting down", workerID)
+					return
+				default:
+				}
+			}
+		}(i)
+	}
+
+	<-ctx.Done()
+	klog.Info("Shutting down GV deletion workers")
```

Comment:
nit: I found it the way these lines combine with `defer h.gvDeletionQueue.ShutDown()` tricky to read.

Maybe above do:

```go
defer func() {
  klog.Info("Shutting down GV deletion workers")
  h.gvDeletionQueue.ShutDown()
}()
```

And then here do:

```
<-ctx.Done() // Wait till context is done to call deferred shutdown function
```

?

I'm assuming that we want to call shutdown immediately if there is a panic raised in this function..